x-pack/plugin/esql/arrow/src/main/java/org/elasticsearch/xpack/esql/arrow/BlockConverter.java

```diff
@@ -0,0 +1,477 @@
+                return length;
+            });
+        }
+
+        private static int offsetVectorLength(BytesRefBlock block) {
+            return Integer.BYTES * (block.getPositionCount() + 1);
+        }
+
+        private int dataVectorLength(BytesRefBlock block) {
+            if (block.areAllValuesNull()) {
+                return 0;
+            }
+
+            // TODO we can probably get the length from the vector without all this sum
+
+            int length = 0;
+            BytesRef scratch = new BytesRef();
+            for (int i = 0; i < block.getPositionCount(); i++) {
+                BytesRef v = block.getBytesRef(i, scratch);
+                length += v.length;
+            }
+            return length;
+        }
+    }
+
+    /**
+     * Conversion of ByteRefs where each value is itself converted to a different format.
+     */
+    public static class TransformedBytesRef extends BytesRefConverter {
+
+        private final BiFunction<BytesRef, BytesRef, BytesRef> valueConverter;
+
+        /**
+         *
+         * @param esqlType ESQL type name
+         * @param minorType Arrow type
+         * @param valueConverter a function that takes (value, scratch) input parameters and returns the transformed value
+         */
+        public TransformedBytesRef(String esqlType, Types.MinorType minorType, BiFunction<BytesRef, BytesRef, BytesRef> valueConverter) {
+            super(esqlType, minorType);
+            this.valueConverter = valueConverter;
+        }
+
+        @Override
+        public void convert(Block b, List<ArrowBuf> bufs, List<BufWriter> bufWriters) {
+            BytesRefBlock block = castBlock(b, BytesRefBlock.class);
+            BytesRefBlock transformed = transformValues(block);
+            super.convert(transformed, bufs, bufWriters);
+        }
+
+        /**
+         * Creates a new BytesRefBlock by applying the value converter to each non null and non empty value
+         */
+        private BytesRefBlock transformValues(BytesRefBlock block) {
+            BlockFactory blockFactory = BlockFactory.getInstance(
+                // FIXME: are these parameters ok here?
+                // FIXME: should the resulting block be closed after use?
+                new NoopCircuitBreaker("esql-arrow"),
+                BigArrays.NON_RECYCLING_INSTANCE
+            );
+
+            try (BytesRefBlock.Builder builder = blockFactory.newBytesRefBlockBuilder(block.getPositionCount())) {
+                BytesRef scratch = new BytesRef();
+                for (int i = 0; i < block.getPositionCount(); i++) {
+                    if (block.isNull(i)) {
+                        builder.appendNull();
+                    } else {
+                        BytesRef bytes = block.getBytesRef(i, scratch);
+                        if (bytes.length != 0) {
+                            bytes = valueConverter.apply(bytes, scratch);
+                        }
+                        builder.appendBytesRef(bytes);
```

Comment:
You are probably better off copying the bytes to a single `BytesRefBuilder` scratch and then passing *that* to the converter to modify in place. That way it doesn't have to allocate the result.

One interesting thing I've realized is that in most cases it's safe to just directly modify the returned `bytes` but not in all cases. Which is kind of lame.